lib/mongo/retryable.rb

```diff
@@ -206,8 +208,10 @@ def write_with_retry(session, write_concern, ending_transaction = false, &block)
       # failed retry_write_allowed? check.
 
       server = select_server(cluster, ServerSelector.primary, session)
+      connection = server.check_out_connection
 
-      unless ending_transaction || server.retry_writes?
+      unless ending_transaction || connection.retry_writes?
+        server.check_in_connection(connection)
         return legacy_write_with_retry(server, session, &block)
```

Comment:
I spent a lot of time thinking about the best way to do this. I wanted to avoid a scenario where `legacy_write_with_retry` is called from within a `server.with_connection` block, causing two connections to be checked out of the same server's pool. I also thought of using the block API but then using `break` or raising an error to get out of the block and then call `legacy_write_with_retry`, but when I wrote that code it was very hard to read. I'd be open to any guidance here.